### ISSUE TITLE
go_repository is very slow for some packages

### ISSUE THREAD

**Issue by arjantop-cai** [bazelbuild/bazel-gazelle#902]
### What version of rules_go are you using?
v0.24.1

### What version of gazelle are you using?
v0.22.0

### What version of Bazel are you using?
3.5.0

### Does this issue reproduce with the latest releases of all the above?
The versions used are latest

### What operating system and processor architecture are you using?
MacOS/amd64

### What did you do?
Generate repositories.bzl with gazelle from go.mod
Run `bazel build //...`

### What did you expect to see?
Go repositories to be fetched quickly

### What did you see instead?
Package github.com/docker/docker needs over 90s to be fetched

<img width="1288" alt="Screenshot 2020-09-08 at 12 10 59" src="https://user-images.githubusercontent.com/66239874/92464495-7f9ff300-f1cd-11ea-92ac-11f4d7b06b93.png">

**Comment by asv** [comment]
i'm guess it's because of the immense **vendor** folder (https://github.com/moby/moby/tree/master/vendor).

**Comment by dirkboye** [comment]
@jayconrod after some debugging I think i've found the problem:

file `repo/remote.go`

```
cmd = exec.Command(goTool, "list", "-find", "-f", "{{.Module.Path}}", "--", importPath)
```

This line is called a LOT of times when discovering repos for `github.com/docker/docker` for example.
Some slow examples for which this line is called:
```
github.com/moby/buildkit/cache
github.com/moby/buildkit/util/resolver
cloud.google.com/go/logging
```

```
$ time A=$(mktemp -d) GO111MODULE="on" GOMODCACHE=$A go list -find -f {{.Module.Path}} -- cloud.google.com/go/logging; sudo rm -rf $A

go: downloading cloud.google.com/go v0.57.0
go: finding module for package cloud.google.com/go/logging
go: downloading cloud.google.com/go/logging v1.1.2
go: downloading cloud.google.com/go v0.75.0
go: found cloud.google.com/go/logging in cloud.google.com/go/logging v1.1.2
go: downloading cloud.google.com/go v0.71.0
go: downloading google.golang.org/genproto v0.0.0-20201106154455-f9bfe239b0ba
go: downloading google.golang.org/api v0.35.0
go: downloading github.com/golang/protobuf v1.4.3
go: downloading google.golang.org/protobuf v1.25.0
go: downloading google.golang.org/grpc v1.33.2
go: downloading golang.org/x/net v0.0.0-20201026091529-146b70c837a4
go: downloading github.com/googleapis/gax-go v2.0.2+incompatible
go: downloading github.com/googleapis/gax-go/v2 v2.0.5
go: downloading golang.org/x/sys v0.0.0-20201013081832-0aaa2718063a
go: downloading golang.org/x/text v0.3.3
go: downloading golang.org/x/sync v0.0.0-20201020160332-67f06af15bc9
go: downloading golang.org/x/oauth2 v0.0.0-20200902213428-5d25da1a8d43
go: downloading go.opencensus.io v0.22.5
go: downloading github.com/google/go-cmp v0.5.2
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
cloud.google.com/go/logging

real	0m9.422s
```

it seems the -find is still downloading all dependencies?

All these calls to `go list -find` sum up to ~100seconds for go_repository(github.com/docker/docker)

**Comment by tommyknows** [comment]
> fetch_repo itself is fast. But the generation of BUILD files for the entire contents of the vendor folder...

I don't think this is correct - when downloading in module mode (e.g. with a sum and version specified in `go_repository`), the `vendor` folder [is deleted right after the download](https://github.com/bazelbuild/bazel-gazelle/issues/547#issuecomment-502864321) (I think the Go tool is doing this, not Gazelle).

However, the general issue still stands and the findings of @dirkboye are correct, `go list` appears to be the culprit. 

We have a similar issue with that behaviour. We're also importing `github.com/docker/docker`, but after fetching the repository, I get issues from transitive dependencies (mostly other `docker` / `moby` projects like `libnetwork`) about the `Sirupsen` / `sirupsen` import path. I tried various fixes, nothing worked so far.
The only thing that did is changing the `go_repository` of `com_github_docker_docker` to download through git (by specifying the commit instead of the version & sum). Gazelle then appears to be honoring the vendor directory.

With this, fetching the docker repository takes ~25s which is probably network-bound.

**Comment by lenage** [comment]
@arjantop-cai just fix sloooow this issue by adding cache when fetch modPath via `go get`,  it's 100+ faster in my project(which has 2000+ importPath)

see commit https://github.com/lenage/bazel-gazelle/commit/c3bc46379ad5c8be93f345d26338a1259a6e0549

**Comment by blico** [comment]
Looks related to #602. 

A list of all known repositories derived from the main repo's WORKSPACE is passed to every `go_repository` rule. Gazelle will then only call out to the network (via `go list`) to resolve imports that do not have a prefix matching any of the passed in repos. However, these network calls are unnecessary because we only should need to resolve repo roots for dependencies that will be built by the main repo.

It appears `github.com/docker/docker` has many cases of references that are not needed by the main repo but are needed internally, resulting in many slow and unnecessary calls to `go list`.

Two potential workarounds would be to pass in `-known_import` and `-exclude` arguments to the the `go_repository` rule via `build_extra_args`. `-known_import` args can tell Gazelle about repos it doesn't know about (e.g. `github.com/moby/buildkit`) to prevent it from going out to the network to resolve them. `-exclude` args can prevent Gazelle from reading parts of the repository that are not needed by them main repo.
